### PR TITLE
Undefined vars as null

### PR DESCRIPTION
**Pull request by terry-xiaoyu**
Fixes https://emqx.atlassian.net/browse/EMQX-12816

Release version: e5.8.1

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [x] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [x] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

**Review by qzhuyan — DISMISSED** [review]

File: apps/emqx_rule_engine/src/emqx_rule_engine_schema.erl [hunk 1]
```diff
@@ -295,6 +295,14 @@ rule_engine_settings() ->
                     deprecated => {since, "v5.0.22"},
                     importance => ?IMPORTANCE_HIDDEN
                 }
+            )},
+        {sql_actions_undefined_vars_as_null,
```
Comment (zmstone) [inline apps/emqx_rule_engine/src/emqx_rule_engine_schema.erl:299]: This impacts ALL rules.
Maybe should be a config for each rule.

Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_engine_schema.erl:299]: We do want it impacts to all `actions` that are able to insert nullable values into databases.

Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_engine_schema.erl]: Renamed to `db_actions_undefined_vars_as_null`.

File: changes/ee/feat-13861.en.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
+A new configuration item `rule_engine.db_actions_undefined_vars_as_null` has been added to the rule engine to ensure that when performing rule databases actions, undefined variables are treated as NULL.
+
+The following actions and SQL functions are affected by this configuration item:
+
+Actions:
+- MySQL
+- PostgreSQL
```
Comment (keynslug) [inline changes/ee/feat-13861.en.md:7]: Nit. It is actually unclear to me how it affects Postrgres connector, does not look like there would be any difference.

File: apps/emqx_rule_engine/src/emqx_rule_engine_schema.erl [hunk 3]
```diff
@@ -295,6 +295,14 @@ rule_engine_settings() ->
                     deprecated => {since, "v5.0.22"},
                     importance => ?IMPORTANCE_HIDDEN
                 }
+            )},
+        {db_actions_undefined_vars_as_null,
```
Comment (keynslug) [inline apps/emqx_rule_engine/src/emqx_rule_engine_schema.erl:299]: Q. Won't it be more natural to have this parameter closer to each action's schema, for example somewhere [here](https://github.com/emqx/emqx/blob/46877e979b638ae17ec79464e7af0100ef4e0b72/apps/emqx_connector/src/emqx_connector_schema_lib.erl#L56-L62)? Right now it looks too broad and a bit out of place.

File: apps/emqx_utils/src/emqx_placeholder.erl [hunk 4]
```diff
@@ -238,6 +253,14 @@ proc_tmpl_deep({tmpl, Tokens}, Data, Opts) ->
 proc_tmpl_deep({tuple, Elements}, Data, Opts) ->
     list_to_tuple([proc_tmpl_deep(El, Data, Opts) || El <- Elements]).
 
+nullable_bin(undefined) -> <<"null">>;
+nullable_bin(null) -> <<"null">>;
+nullable_bin(Var) -> bin(Var).
+
+nullable_bin_dynamo(undefined) -> {null, true};
+nullable_bin_dynamo(null) -> {null, true};
+nullable_bin_dynamo(Var) -> bin(Var).
```
[the remark this answers is not in the document]
Comment (terry-xiaoyu) [reply apps/emqx_utils/src/emqx_placeholder.erl:262]: fixed

File: changes/ee/feat-13861.en.md (hunk 2 above)
Comment (terry-xiaoyu) [reply changes/ee/feat-13861.en.md:7]: fixed

File: apps/emqx_rule_engine/src/emqx_rule_engine_schema.erl (hunk 3 above)
Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_engine_schema.erl:299]: changed

File: apps/emqx_utils/src/emqx_utils_sql.erl [hunk 5]
```diff
@@ -135,6 +136,14 @@ to_sql_string(Term, #{escaping := cql}) ->
...
     maybe_escape(Term, fun escape_sql/1).
 
+-spec sqlstr_opts(map()) -> map().
+sqlstr_opts(Opts) ->
+    Path = [emqx_rule_engine_schema:namespace(), db_actions_undefined_vars_as_null],
+    case emqx:get_config(Path, false) of
+        false -> Opts#{undefined => <<"undefined">>};
+        true -> Opts
+    end.
+
```
[the remark this answers is not in the document]
Comment (terry-xiaoyu) [reply apps/emqx_utils/src/emqx_utils_sql.erl]: fixed

File: apps/emqx_rule_engine/src/emqx_rule_engine_schema.erl (hunk 1 above)
Comment (terry-xiaoyu) [reply apps/emqx_rule_engine/src/emqx_rule_engine_schema.erl]: refactored to add dedicate options for each action

File: apps/emqx_utils/src/emqx_placeholder.erl [hunk 6]
```diff
@@ -254,14 +271,26 @@ bin(Val) -> emqx_utils_conv:bin(Val).
 quote_sql(Str) ->
     emqx_utils_sql:to_sql_string(Str, #{escaping => sql, undefined => <<"undefined">>}).
 
+-spec quote_sql2(_Value) -> iolist().
+quote_sql2(Str) ->
+    emqx_utils_sql:to_sql_string(Str, #{escaping => sql}).
+
```
Comment (keynslug) [inline apps/emqx_utils/src/emqx_placeholder.erl:277]: Nit. Naming is far from obvious, maybe `quote_sql_bwcompat/1` for the former and `quote_sql/1` for the latter?

**Review by keynslug — COMMENTED** [review]
Great work. 👍🏼 Just a small nit from me.

**Review by zmstone — CHANGES_REQUESTED** [review]
test coverage

**Review by zmstone — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-4415)

**Pull request merged**